### ISSUE TITLE
Module not found: Can't resolve '@babel/runtime/helpers/esm/extends'

### ISSUE THREAD

**Issue by HeleneBigSofa** [babel/babel#9746]
## Bug Report

**Current Behavior**
I created an app within a monorepo (Lerna) using yarn workspace and create-react-app ("react-scripts": "2.0.0-next.3e165448").

The architecture of the app is as follow:
- node_modules
      - @babel
            - core
            - runtime
            - etc..
      - other modules...
- package1 (shared components)
      - src
      - package.json
- package2 (the app consuming the shared components)
      - src
      - package.json
- package.json

**Input Code**
Calling
   ```import _extends from '@babel/runtime/helpers/esm/extends';```
from package2/index.js (or in any dependencies of package1) results in this error:
`Module not found: Can't resolve '@babel/runtime/helpers/esm/extends'`

Calling any other modules available under /node_modules works fine, even from @babel/core for instance:
```import _default from '@babel/core/lib/tools/build-external-helpers';```
works correctly.

Please note, if I replace @babel/runtime/... with the absolute path it would work.

**Expected behavior/code**
Need to be able to call @babel/runtime/...

**Babel Configuration (.babelrc, package.json, cli command)**
In the package.json root:

```
"babel": {
    "presets": [
      [
        "@babel/preset-env"
      ]
    ],
    "plugins": [
      "@babel/plugin-transform-runtime"
    ]
  },
```

**Environment**
- Babel version(s): @babel/runtime 7.4.2, @babel/core 7.4.0
- Node/npm version: Node v10.15.2
- OS:  OS High Sierra 10.13.6
- Monorepo:  yes, Lerna

**Comment by nicolo-ribaudo** [comment]
Can you post all your package.jsons?

**Comment by HeleneBigSofa** [comment]
The package.jsons:
In the root 
```
{
  "name": "monorepo",
  "private": true,
  "workspaces": {
    "packages": [
      "package1/",
      "package2/",
      "package2/src/**"
    ],
    "nohoist": [
      "**/eslint",
      "react-scripts",
      "react-scripts/**",
      "**/jest/**"
    ]
  },
  "jest": {
    "testPathIgnorePatterns": [
      "<rootDir>/node_modules/",
      "<rootDir>/spa/"
    ],
    "projects": [
      "<rootDir>/playbook/*"
    ]
  },
  "husky": {
    "hooks": {
      "pre-commit": "lint-staged"
    }
  },
  "dependencies": {
    "apollo-boost": "^0.1.16",
    "graphql": "^14.0.2",
    "react-apollo": "^2.2.4"
  },
  "devDependencies": {
    "@babel/cli": "^7.1.2",
    "@babel/core": "^7.1.2",
    "@babel/preset-env": "^7.1.0",
     "@babel/runtime": "^7.4.2",
     "@babel/plugin-transform-runtime": "^7.4.0",
    "babel-eslint": "9",
    "eslint": "^4.17.0",
    "eslint-config-airbnb": "^16.1.0",
    "eslint-config-prettier": "^2.9.0",
    "eslint-plugin-import": "^2.8.0",
    "eslint-plugin-jsx-a11y": "^6.0.3",
    "eslint-plugin-prettier": "^2.6.2",
    "eslint-plugin-react": "^7.6.1",
    "husky": "^1.0.0-rc.13",
    "lerna": "^3.4.3",
    "lint-staged": "^7.2.2",
    "prettier": "^1.14.2",
    "prettier-eslint": "^8.8.1",
    "prettier-eslint-cli": "^4.7.0"
  },
  "lint-staged": {
    "./**/*.{js,jsx}": [
      "yarn eslint",
      "git add"
    ]
  },
  "babel": {
    "presets": [
      [
        "@babel/preset-env"
      ]
    ],
    "plugins": [
      "@babel/plugin-transform-runtime"
    ]
  },
}

```

In package1: 
```
{
  "name": "package2",
  "version": "1.0.30",
  "private": true,
  "dependencies": {
    "@date-io/moment": "^1.0.2",
    "@material-ui/core": "^3.3.1",
    "@material-ui/icons": "^3.0.1",
    "@material-ui/lab": "^3.0.0-alpha.25",
    "@storybook/react": "^3.4.8",
    "apollo-storybook-react": "^0.1.8",
    "autosuggest-highlight": "3.1.1",
    "axios": "^0.18.0",
    "date-fns": "2.0.0-alpha.27",
    "faker": "^4.1.0",
    "graphql": "^14.0.2",
    "graphql-iso-date": "^3.6.1",
    "graphql-tag": "^2.9.2",
    "graphql-tools": "^3.1.1",
    "jest-dom": "^1.12.0",
    "jest-styled-components": "^6.3.1",
    "lodash.defaults": "^4.2.0",
    "lodash.isequal": "^4.5.0",
    "material-ui": "^0.20.2",
    "material-ui-chip-input": "^1.0.0-beta.14",
    "material-ui-pickers": "^2.2.1",
    "material-ui-slider": "^3.0.0",
    "moment": "^2.24.0",
    "react": "^16.4.1",
    "react-adopt": "^0.6.0",
    "react-autosuggest": "^9.4.3",
    "react-dom": "^16.4.1",
    "react-dropzone": "^5.1.0",
    "react-router": "^5.0.0",
    "react-router-dom": "^5.0.0",
    "react-scripts": "2.0.0-next.3e165448",
    "react-testing-library": "^5.0.0",
    "storybook": "^1.0.0",
    "storybook-addon-material-ui": "^0.9.0-alpha.16",
    "storybook-react-router": "^1.0.1",
    "storybook-readme": "4.0.0-beta1",
    "styled-components": "^3.4.5",
    "video.js": "^7.3.0"
  },
  "devDependencies": {
    "@sambego/storybook-state": "^1.3.1",
    "@storybook/addon-a11y": "^3.4.10",
    "@storybook/addon-actions": "^3.4.11",
    "@storybook/addon-info": "^3.4.10",
    "@storybook/addon-jest": "^3.4.10",
    "@storybook/addon-knobs": "^3.4.10",
    "@storybook/addon-links": "^3.4.11",
    "@welldone-software/why-did-you-render": "^2.5.0",
    "apollo-storybook-react": "^0.1.8",
    "axios-mock-adapter": "^1.16.0"
  },
  "jest": {
    "collectCoverageFrom": [
      "src/**/*.{js,jsx}",
      "!/node_modules/",
      "!src/index.js",
      "!src/registerServiceWorker.js",
      "!src/**/stories.{js,jsx}",
      "!src/api/**/*.{js,jsx}"
    ]
  }
}

```

In package2:

```
{
  "name": "spa",
  "version": "1.0.30",
  "private": true,
  "dependencies": {
    "apollo-boost": "^0.1.12",
    "graphql": "^0.13.2",
    "react": "^16.4.1",
    "react-apollo": "^2.1.9",
    "react-dom": "^16.4.1",
    "react-scripts": "2.0.0-next.3e165448",
    "source-map-explorer": "^1.6.0"
  },
  "scripts": {
    "start": "PORT=3003 react-scripts start",
    "analyze": "source-map-explorer build/static/js/vendors.*",
    "build": "react-scripts build",
    "test": "react-scripts test --env=jsdom",
    "eject": "react-scripts eject",
    "cypress:open": "cypress open",
    "cypress:run": "cypress run"
  },
  "browserslist": {
    "development": [
      "last 2 chrome versions",
      "last 2 firefox versions",
      "last 2 edge versions"
    ],
    "production": [
      ">0.25%",
      "not op_mini all",
      "ie 11"
    ]
  },
  "devDependencies": {
    "cypress": "^3.1.5",
  }
}

```

**Comment by nicolo-ribaudo** [comment]
I think that the problem is that lerna assumes that you aren't using `@babel/runtime` in `package2`, because you didn't specify it in its dependencies. For this reason, if any dependency of `package2` depends on `@babel/runtime`, lerna thinks that it can safely hoist that dependency's `@babel/runtime` to `package2/node_modules`. If this hoisted version is an old version without the `extends` helper, it will throw.

I would try adding `@babel/runtime` to `paackage2` (and `package1`, if needed): even when using a monorepo, every runtime dependency should be listed.

Your package.json also has two other problems:
1) `@babel/runtime` is used at runtime (not at build time) so it should be in `dependencies`, not `devDependencies`
2) (maybe this is intentional) The Babel config defined in the root package won't be applied to package1/2. If you want it to be, you should use `babel.config.js` instead: https://babeljs.io/docs/en/config-files

**Comment by HeleneBigSofa** [comment]
Thank you for your reply and solutions Nicolo, well appreciated.
I already tried to install @babel/runtime as a direct dependency of package2, but let me try again, I might have missed something. I also applied a "nohoist" of @babel/runtime in the root package.json but it did not change the issue unfortunately. I'll try point 1 and 2 and come back to you.

**Comment by nicolo-ribaudo** [comment]
Point 1 and 2 are just "details", I don't think that they will solve your problem.
What version does `package2/node_modules/@babel/runtime/package.json` say?

**Comment by HeleneBigSofa** [comment]
it says: ``` "version": "7.4.2"```

**Comment by nicolo-ribaudo** [comment]
And doesn't `helpers/esm/extends.js` exist inside that package?

**Comment by HeleneBigSofa** [comment]
Yes it is definitely there. And if I import it with: 
```
import _extends from '../../node_modules/@babel/runtime/helpers/esm/extends';
``` 
it's found.

**Comment by nicolo-ribaudo** [comment]
That's really strange :thinking:
Could you setup a repository which reproduces the issue?

**Comment by HeleneBigSofa** [comment]
I set up a repo that reproduce the error: https://github.com/htmlene/monorepo

When I first raised the issue, I gave a simplistic example by calling directly @babel/runtime in the package2. The monorepo example is closer to the reality as @babel/runtime is required by a dependency "react-router" (+history/react-router-dom) - I hope it makes sense, let me know if you have any questions.

**Comment by nicolo-ribaudo** [comment]
I will check it later today

**Comment by HeleneBigSofa** [comment]
Thank you Nicolo, it's very kind of you

**Comment by nicolo-ribaudo** [comment]
I think that this might be a bug with `react-scripts` (or with the bundler it internally use; I don't know which it is). The Babel package is definetly there (if you go in `monorepo/node_modules/history/esm`, open a node shell and run `require.resolve("@babel/runtime/helpers/esm/extends")` it clearly show that that module can be resolved).
I also tried to disable yarn workspaces and use the lerna implementation of workspaces, but it didn't help.

**Comment by HeleneBigSofa** [comment]
Thank you Nicolo. Do you think reverting to the previous version of @babel we were on would solve the issue?

**Comment by nicolo-ribaudo** [comment]
I'm don't think so, but it might be worth trying

**Comment by HeleneBigSofa** [comment]
We have finally managed to solve our issue, using react-app-rewired, customize-cra and a custom config. 
Thank your for your help Nicolo!

**Comment by yanyu0517** [comment]
We have met the same problem. Was there news about this ?

**Comment by nelhu** [comment]
I have the same problem and did not resolve yet

**Comment by fgiarritiello** [comment]
@HeleneBigSofa Do you mind sharing the solution? Thanks

**Comment by HeleneBigSofa** [comment]
We just upgraded react-script to version 2.1.8

**Comment by fgiarritiello** [comment]
@HeleneBigSofa thank you. Sorry but I have a different question then: how did you manage to get the monorepo to work if react-script dropped its support after the version you were using?

**Comment by ivan-kleshnin** [comment]
Same problem with MDX + Next. Tried a few recommended workaround – nothing helps.

**Comment by crubier** [comment]
Same problem when building a relatively simple package with rollup... No solution works so far..

**Comment by thecpdubguy** [comment]
Having this issue as well with react-scripts 3.0.1

**Comment by nicolo-ribaudo** [comment]
Is the `@babel/runtime/helpers/esm/objectSpread` file there?

**Comment by gilbertlucas46** [comment]
I fixed mine by

1. Delete package-lock.json
2. Delete node-modules
3. npm install
